Axi pulse capture driver

nunojsa opened the pull request:
This driver enables support for the axi_pulse_capture driver which
controls the state of a laser driver and a pulse generator signal (pwm).
The driver register an iio device with an IIO_ALTVOLTAGE channel type
(since there was no better fit). Also note, that it was decided to have
this driver under adc directory since the pwm signal generated by the
pulse generator controls an adc capture start signal (and no other place
seemed to be suitable).

Signed-off-by: Nuno Sa <Nuno.Sa@analog.com>

mirceacaprioru commented on Documentation/devicetree/bindings/iio/adc/adi,axi-pulse-capture.txt:
```diff
@@ -0,0 +1,27 @@
+ Analog Device AXI Pulse and Laser driver controller
```
dt-bindings should be in their own commit.

mirceacaprioru replied:
Also there is a new requirement for bindings to be in machine readable YAML format. This is a transition from *.txt format.

nunojsa replied:
Updated to it's own commit. I was not aware of the new YAML requirement. But since this driver is not intended to be mainline (at least in the near future) i will keep the *.txt file. But i will keep this in mind for the future...

mirceacaprioru commented on arch/arm/boot/dts/zynq-zc706-adv7511-fmclidar1.dts:
```diff
@@ -39,9 +39,19 @@
 		#size-cells = <0>;
```
I would also put the dts file in its own commit.

nunojsa replied:
Done

mirceacaprioru commented on drivers/iio/adc/axi_pulse_capture.c:
```diff
@@ -0,0 +1,454 @@
+#include <linux/io.h>
+#include <linux/interrupt.h>
+#include <linux/fpga/adi-axi-common.h>
+#include <linux/clk.h>
```
alphabetical order is preferred for includes

mirceacaprioru commented on drivers/iio/adc/axi_pulse_capture.c:
```diff
@@ -0,0 +1,454 @@
+	pr_debug("Set value (%u), laser EXT Clk rate (%lu)\n",
+		cnt, clk_rate);
+
+	pulse_capture_iowrite(pulse, reg, cnt);
```
doesn't pulse_capture_iowrite() require a lock procedure too ?

nunojsa replied:
I don't think so, we just have to synchronize the IRQ logic with the IIO_CHAN_INFO_ENABLE logic or when actually enabling the device or configuration register. So that, we are not left in a state were we think the device is enabled when it is not.

mirceacaprioru commented on drivers/iio/adc/axi_pulse_capture.c:
```diff
@@ -0,0 +1,454 @@
+
+	switch (mask) {
+	case IIO_CHAN_INFO_ENABLE:
+		*val = pulse_capture_ioread(pulse, ADI_REG_DRIVER_ENABLE);
```
would a lock be required here ?

nunojsa replied:
I don't think so. We are just reading a register from the device. I'm not seeing what could we gain from a lock. If someone else is writing on the same register we will get the value from whoever accessed the register first.

commodo replied:
I would agree with Mircea here a bit.
The thing is, that lock use is inconsistent in this code.
[At this point](https://github.com/analogdevicesinc/linux/pull/352/commits/ac300f2b613974a2eb348b73d781b19f947dfbff#diff-97810dfba0168edc6da6238b6a563fa4R85) this register is being accessed with a lock, while here it's not.

There is some lack of consistency about this locking.
So then the question is: does this need locking or not ?
Whether the answer is yes or no, it does not matter. Consistency matters a bit more (in this case).

nunojsa replied:
Here I don't have a lock because I just don't think we need one. So we don't need to set our current task to sleep just to read this register. I placed the lock in logic which I actually wanted to synchronize and, if not, might leave the device in a state not coherent. For instance having the lock between the IRQ thread and the `IIO_CHAN_INFO_ENABLE` is important to synch the `otw` boolean. If we get an otw event we want to disable the device and forbid the user to enable it at the same time (until we leave this state). Now, if at the same time of the IRQ for instance we are writing (in some attribute other than the ENABLE) or reading  some other register  i don't think it's an issue and i don't think that can leave the device in an incoherent state. The lock in these case would be just overhead i believe...

commodo replied:
i'm a bit un-decided about the lock here;
typically if we would read data in this callback, then we wouldn't lock;
but this deals more with reading system-state, which ideally should be consistent when being read, and no-locking here looks suspicious [if presenting this to upstream usually]

but we won't upstream this anytime soon; so let's leave it like this for now;
if we get a commit [later] adding a lock, i will be amused :)

nunojsa replied:
I think even with the lock it's always about timing. Let's say we are reading this register while, concurrently, an IRQ comes with an OTW enter event  (disabling the device). If the IRQ get's the lock first, then this read will already have the updated value. But, if the reader get's the lock first than the IRQ, then we will get a value which is about to change. The same goes without the lock, depending on whoever writes/reads the register first (`iowrite()/ioread()` guarantee coherency here using barriers). Either way, as you said, let's leave it like this for now and if we need to change it for mainlining the driver we do it :).

mirceacaprioru commented on drivers/iio/adc/axi_pulse_capture.c:
```diff
@@ -0,0 +1,454 @@
+					unsigned int writeval,
+					unsigned int *readval)
+{
+	struct axi_pulse_capture *pulse = iio_priv(indio_dev);
```
we could place a lock inside pulse_capture_iowrite structure and take care of it inside pulse_capture_iowrite() and pulse_capture_ioread(). All these locks are hard to follow outside of their scope.

mirceacaprioru commented on drivers/iio/adc/axi_pulse_capture.c:
```diff
@@ -0,0 +1,453 @@
+	struct axi_pulse_capture *pulse = iio_priv(indio_dev);
+	u32 irq_pending = 0;
+
+	mutex_lock(&indio_dev->mlock);
```
i think you will need a spin_lock_irqsave() to be able to restore the interrupt to its previous state while you are performing critical tasks. Also you are in a threaded IRQ here.

nunojsa replied:
I'm not seeing why i would need `spin_lock_irqsave()`. I don't think it's necessary to disable the interrupts at this point. Since it's threaded IRQ we can sleep and use the mutex... And AFAIK the operation does not need to be atomic...

mirceacaprioru replied:
you're right. I assumed wrong, this is the actual handler for the irq, not some other function where you would save the state of the irq.

commodo replied:
this one place i would definitely avoid using `&indio_dev->mlock` and define a specific driver lock;
it's really hard to tell in which state the IIO framework is to be able to say "yes-this-won't-ever-deadlock"

nunojsa replied:
I agree with the own lock and i will change it. But i think this is actually the one place where it is safer :D (but i might be forgetting something). This is called in their own kernel thread context, after the top half interrupts is served it just return `IRQ_WAKE_THREAD` which will schedule some kernel thread to serve this. Now, in user context (calling stuff like `->read_raw()`) that might become an issue in the future and lead to possible deadlocks...

mirceacaprioru commented on drivers/iio/adc/axi_pulse_capture.c:
```diff
@@ -0,0 +1,453 @@
+			pulse->otw = true;
+		}
+	}
+	mutex_unlock(&indio_dev->mlock);
```
following previous comment spin_unlock_irqrestore()

stefpopa commented on drivers/iio/adc/axi_pulse_capture.c:
```diff
@@ -0,0 +1,453 @@
+		if (ret < 0) {
+			dev_warn(dev, "Could not get pulse-frequency (%d)\n",
+				 ret);
+			return;
```
`return ret; `?

nunojsa replied:
This API is not returning (void). And the idea is to have some default values which come from dts. So, I don't think we should fail the `probe()` just because we fail to do the default configurations... Should we actually return the error and fail the `probe()` (seemed a little extreme to me)?

stefpopa replied:
Ok, then these properties should be optional. In the dt-binding they are stated as "Required properties".

nunojsa replied:
Ok, it makes sense. I will also update that

nunojsa replied:
Now, if one of these properties is not set, the driver won't enable the device by default. It can still be done via sysfs/iio. Hmm but i guess it makes sense to have some default values in case these (or one them) are not set.

stefpopa commented on Documentation/devicetree/bindings/iio/adc/adi,axi-pulse-capture.txt:
```diff
@@ -0,0 +1,27 @@
+	//adi,laser-driver-disabled;
+
+	adi,pulse-width-ns = <100>;
+	adi,pulse-frequency = <1000000>;
```
`adi,pulse-frequency-hz`

nunojsa replied:
Makes sense...

stefpopa commented on arch/arm/boot/dts/zynq-zc706-adv7511-fmclidar1.dts:
```diff
@@ -39,9 +39,19 @@
+		//adi,laser-driver-disabled;
+
+		adi,pulse-width-ns = <100>;
+		adi,pulse-frequency = <1000000>;
```
Actually, to me it looks that these properties do not belong in the device tree, since they do not define the hardware and are just used for configuration. Moreover, they are optional and initialized with 0 if the field is left empty. Why not use IIO attributes? This way we can also modify them on the fly.
For frequency, `IIO_CHAN_INFO_SAMP_FREQ` can be used and for the others we can create some custom attributes.

nunojsa replied:
```
static const struct iio_chan_spec_ext_info ext_info = {
	.name = "pulse_width_ns",
	.read = axi_pulse_capture_read_pulse_w,
	.write = axi_pulse_capture_write_pulse_w,
};

static const struct iio_chan_spec axi_pulse_capture_channels[] = {
	{
		.type = IIO_ALTVOLTAGE,
		.info_mask_separate = BIT(IIO_CHAN_INFO_ENABLE) |
				BIT(IIO_CHAN_INFO_FREQUENCY),
		.channel = 0,
		.indexed = 1,
		.output = 1,
		.ext_info = &ext_info,
	},
};
```
I'm already using IIO attributes to change this values on the fly. The idea was just to have some initial/default values that we could change without having to recompile the driver. But i can just drop these properties and use some hard-coded default values in the driver... I just added these properties in the device tree because I saw we are doing it for other devices already (but i agree this is not describing hardware)...

stefpopa replied:
Aaah, ok! I haven't checked if you are using IIO attributes :smiley: ! I think that you are right and it's a better approach to just use some hard-coded default values.

commodo commented on drivers/iio/adc/axi_pulse_capture.c:
```diff
@@ -0,0 +1,416 @@
+	}
+
+	dev_dbg(indio_dev->dev.parent, "Set pulse width to %lu\n",
+			(1000000000U/pulse_time));
```
maybe assign this division to a variable, since it's used here in `dev_dbg` & `axi_pulse_capture_params_set()`
or remove  `dev_dbg`  if no longer needed

nunojsa replied:
Agreed

commodo commented on drivers/iio/adc/axi_pulse_capture.c:
```diff
@@ -0,0 +1,416 @@
+		pulse_capture_iowrite(pulse,
+		ADI_REG_CONFIG, PULSE_GEN_LOAD);
+	}
+	mutex_unlock(&indio_dev->mlock);
```
using `indio_dev->mlock` is discouraged ; you should define a new one and initialize on `struct axi_pulse_capture`
`indio_dev->mlock` is typically used at the IIO framework level, and the maintainer has acknowledged that it was misused a bit; 
some old driver still use it, and so it's not too bad if we use it, but going forward i would avoid potential issues (with this lock) that would come from future IIO changes

nunojsa replied:
To be honest, I followed the approach that we have in another driver and din't gave to much thought. But I agree with you and I'm going to instantiate our own lock. If the IIO core changes and starts to lock this mutex before calling our callbacks we can have serious problems...

commodo commented on drivers/iio/adc/axi_pulse_capture.c:
```diff
@@ -0,0 +1,416 @@
+		pulse_capture_iowrite(pulse, reg, writeval);
+	} else {
+		*readval = pulse_capture_ioread(pulse, reg);
+		dev_dbg(indio_dev->dev.parent, "read addr 0x%08X, val=0x%08X\n",
```
generally, if they're no longer needed, i would remove `dev_dbg` code

commodo commented on drivers/iio/adc/axi_pulse_capture.c:
```diff
@@ -0,0 +1,416 @@
+
+	mutex_lock(&indio_dev->mlock);
+	irq_pending = pulse_capture_ioread(pulse, ADI_REG_IRQ_PENDING);
+	dev_info(indio_dev->dev.parent,
```
`dev_info` would spam dmesg with interrupts

nunojsa replied:
Unless we have a lot of state changes that is unlikely. The interrupts come at the positive edge of the signal (leaving over temperature warning) and on the negative edge (entering over temperature warning). For now, i would leave it and if we catch some situation where this can actually happen we can remove it...

commodo replied:
maybe change this `dev_info_ratelimited` then;
if something electrical happens, at least the log is manage-able

nunojsa replied:
I'm not sure we actually need it but I'm ok with making this change just to be super cautious. Moreover, after discussing with @mhennerich, it makes more sense to use `dev_warn_ratelimited`.

stefpopa commented:
Can you please add a changelog after every force push?

stefpopa commented on drivers/iio/adc/axi_pulse_capture.c:
```diff
@@ -0,0 +1,414 @@
+	 size_t len)
+{
+	struct axi_pulse_capture *pulse = iio_priv(indio_dev);
+	int ret = 0;
```
nitpick: there is no need to initialize this variable.

stefpopa commented on drivers/iio/adc/axi_pulse_capture.c:
```diff
@@ -0,0 +1,414 @@
+ * Copyright 2016 Analog Devices Inc.
+ *
+ * Licensed under the GPL-2.
+ */
```
curious: should we use the SPDX identifier for this type of drivers as well?

nunojsa replied:
To be hones, I don't know :smile:

stefpopa replied:
Me neither, but upstream it seems to be the norm. So maybe should we use it as well on all our Linux drivers?

nunojsa replied:
Yes, reading the Documentation seems to be the norm. However, looking for other drivers, it seems that the rule is not being strictly respected. Either way, I will add the  `SPDX` line in the firs line of the file...

stefpopa replied:
You are right that it is not strictly respected, but it is probably because the SPDX identifier is a new thing. It has been introduced more or less one year ago. Upstream there has been some effort to update existing drivers to use it.

stefpopa commented on drivers/iio/adc/axi_pulse_capture.c:
```diff
@@ -0,0 +1,414 @@
+	int ret = 0;
+
+	mutex_lock(&pulse->pulse_lock);
+	if (readval == NULL) {
```
nitpick: `if (!readval)` :smiley:

stefpopa commented on drivers/iio/adc/axi_pulse_capture.c:
```diff
@@ -0,0 +1,414 @@
+	pulse_capture_iowrite(pulse, ADI_REG_IRQ_MASK,
+			  (ADI_IRQ_MASK_OUT_ALL &
+			  ~(ADI_IRQ_SRC_OTW_ENTER | ADI_IRQ_SRC_OTW_EXIT)));
+
```
nitpick: extra line

stefpopa commented on drivers/iio/adc/axi_pulse_capture.c:
```diff
@@ -0,0 +1,414 @@
+	pulse_capture_iowrite(pulse, ADI_REG_DRIVER_ENABLE, LASER_ENABLE);
+	/* All interrupts are disabled on the core at startup.
+	 * Let's only enable the otw irqs.
+	 */
```
multi line comments format is:
```
/* 
 * comment
 */
```

commodo commented on drivers/iio/adc/axi_pulse_capture.c:
```diff
@@ -0,0 +1,414 @@
+		 irq_pending, pulse->otw, pulse->saved_en);
+	if (pulse->otw) {
+		/* we are waiting for an exit transition */
+		if (irq_pending & ADI_IRQ_SRC_OTW_EXIT) {
```
mostly a stylistic comment;
this has a few too many levels of indentation;
i would break this into a separate function (something like `axi_laser_irq_src_otw_exit` or some other name) and call it here;
same for `axi_laser_irq_src_otw_enter`

in such a function you could just do
```
if (!(irq_pending & ADI_IRQ_SRC_OTW_EXIT))
    return;
```

then the rest of the code is 1-level indented

nunojsa replied:
I think the indentation is still not terribly bad here but i will make the change either way :smile:

commodo commented on drivers/iio/adc/axi_pulse_capture.c:
```diff
@@ -0,0 +1,414 @@
+		mutex_unlock(&pulse->pulse_lock);
+		return ret;
+	case IIO_CHAN_INFO_FREQUENCY:
+		return axi_pulse_capture_params_set(pulse,
```
i would put the lock here as well;
mostly for consistency

commodo replied:
wait; nevermind; found the locks inside `axi_pulse_capture_params_set()`

nunojsa replied:
I would prefer not to lock it. In my opinion, if there is no need for the lock, just don't add the extra overhead. Locking an already held mutex is not a lightweight thing (i know we don't really need to be performant in this driver though :smile: ).

commodo commented on drivers/iio/adc/axi_pulse_capture.c:
```diff
@@ -0,0 +1,414 @@
+	switch (mask) {
+	case IIO_CHAN_INFO_ENABLE:
+		mutex_lock(&pulse->pulse_lock);
+		if (!pulse->otw) {
```
stylistic; you could reduce the level of indented code here a bit:
```
if (pulse->otw) {
	dev_warn(indio_dev->dev.parent,
		 "Don't touch the enable register while in over temperature state!\n");
	mutex_unlock(&pulse->pulse_lock);
        return -EPERM;
}
```

nunojsa replied:
Agreed

stefpopa commented:
There is a checkpatch complaint: WARNING: please write a paragraph that describes the config symbol fully
#43: FILE: drivers/iio/adc/Kconfig:374:
+config AXI_PULSE_CAPTURE

nunojsa commented:
Changes in V4:

- Added `SPDX` line;

- Fixed checkpatch.pl warning on Kconfig;

- Reworked some code for stylistic reasons (decrease level of indentation in some places);

- Fixed multi-line comments.

stefpopa commented on drivers/iio/adc/axi_pulse_capture.c:
```diff
@@ -0,0 +1,427 @@
+ *
+ * Copyright 2016 Analog Devices Inc.
+ *
+ * Licensed under the GPL-2.
```
If you added the SPDX license identifier, there is no need for this line.

nunojsa replied:
Obviously :man_facepalming:.

nunojsa commented:
Changes in V5:

- Removed unneeded line in the file license header.

commodo commented on drivers/iio/adc/axi_pulse_capture.c:
```diff
@@ -0,0 +1,425 @@
+	indio_dev->num_channels = ARRAY_SIZE(axi_pulse_capture_channels);
+	indio_dev->info = &axi_pulse_capture_info;
+
+	ret = devm_iio_device_register(&pdev->dev, indio_dev);
```
maybe let's move this `devm_iio_device_register` to be the last [functionally];
so, after `axi_pulse_capture_config` and after `mutex_init`;
after iio_device_register happens, the device should be accessible to userspace, and potentially un-configured, and some bits could be [still] un-initialized;

nunojsa replied:
Agreed

nunojsa commented:
Changes in V6:

- Moved `devm_iio_device_register()` to the end of the `probe()` function;
- Use `dev_warn_ratelimited()` instead of `dev_info()` in the IRQ handler.
